[tests] Skip XslCompiledTransformApi tests on iOS, tvOS, and when testing native AOT

mdh1418 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/75699

In effort to consolidate System.Private.Xml Test projects, the test suite exclusions within https://github.com/dotnet/runtime/blob/main/src/libraries/tests.proj weren't being honored because the suites no longer existed. 

Instead, to skip the tests, `[ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsReflectionEmitSupported))]` is being added at the class level to skip tests where Reflection Emit is needed.

This PR skips tests within XslCompiledTransformAp and XmlSerializerTests

mdh1418 committed:
Skip XslCompiledTransformApi tests on iOS, tvOS, and when testing native AOT

mdh1418 commented:
/azp run runtime-extra-platforms

MichalStrehovsky commented on src/libraries/System.Private.Xml/tests/System.Private.Xml.Tests.csproj:
```diff
@@ -621,7 +621,8 @@
   </ItemGroup>
 
-  <ItemGroup>
+  <ItemGroup Condition="(('$(TargetOS)' != 'tvOS' and '$(TargetOS)' != 'iOS') or '$(RunDisablediOSTests)' == 'true') and
+                        ('$(TestNativeAot)' != 'true' or '$(RunDisabledNativeAotTests)' == 'true')" >
```
It might be better to put `[ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsReflectionEmitSupported))]` on the individual test classes in these files. It's more consistent with how we do things and more futureproof.

mdh1418 committed:
Condition each XslCompiledTransformApi test class on IsReflectionEmitSupported

mdh1418 commented:
/azp run runtime-extra-platforms

stephentoub reviewed (approved):
LGTM if it fixes the problem.

filipnavara commented:
It fixes half of the problem. Other half of the NativeAOT tests still fail.

mdh1418 committed:
[tests] Condition XmlSerializerTests on ReflectionEmit
Tests fail from trimming and MakeGeneric, and XmlSerializer is not supported with NativeAOT

mdh1418 commented:
This PR was initially only targeting XslCompiledTransformApi tests, so I didn't have a look at all the other test suites that were removed and where the respective TestExclusions weren't honored, but I'll look at the failing XmlSerializerTests tests. Would it make sense to make PRs on a test suite by test suite basis, or trying to tackle them all in this one PR?

filipnavara commented:
I don't know what triggered it but my totally unrelated PR started seeing the same errors on the NativeAOT pipelines. In fact I started getting them locally for the `main` branch when building the NativeAOT tests. That's why I started watching this PR since it resolves a sizeable amount of the failures. I appreciate that you started looking into it even if it was not for the same underlying reason.

Btw, you should run the extras pipelines (`/azp run runtime-extra-platforms`) once you are reasonably sure that the tests from the previous run are now marked correctly.

MichalStrehovsky commented:
>  don't know what triggered it but my totally unrelated PR started seeing the same errors on the NativeAOT pipelines

The test projects (mentioned in the diffs to src/libraries/tests.proj in this PR) got deleted and folded into another project that is not excluded. So tests we weren't even compiling before started failing.

filipnavara commented:
> The test projects (mentioned in the diffs to src/libraries/tests.proj in this PR) got deleted and folded into another project that is not excluded. So tests we weren't even compiling before started failing.

Ah, that explains it. I can understand that it was missed as part of a PR if the extras pipelines were not run but I am quite surprised that there was no issue filed for it based on the rolling builds.

stephentoub commented:
https://github.com/dotnet/runtime/issues/75699#issuecomment-1249930017

filipnavara commented:
> [#75699 (comment)](https://github.com/dotnet/runtime/issues/75699#issuecomment-1249930017)

Missed that one (or rather it was not filed yet when I first hit the issue). Thanks.

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 commented:
> It fixes half of the problem. Other half of the NativeAOT tests still fail and need similar treatment.

I looked at the test failures more closely, and it doesn't look obviously System.Reflection.Emit to me.
Seeing errors like `System.InvalidOperationException : There is an error in XML document (2, 14).`

I looked at the comment in tests.proj and it looks like System.XmlSerializer is just not supported on NativeAOT right now.

I also noticed some test failures in two more suites that were removed `System.Xml.Xsl.XslCompiledTransformApi.Tests.csproj` however there doesn't seem to be a test suite skip in tests.proj, so not sure how these were skipped in the past. They were failing with
```
System.Xml.XmlSchemaValidatorApiTests.VerifyException : Error while loading assembly
   at System.Xml.XmlSchemaValidatorApiTests.ExceptionVerifier..ctor(String assemblyName, ExceptionVerificationFlags flags, ITestOutputHelper output) + 0x5fd
   at System.Xml.XmlSchemaValidatorApiTests.TCInitialize..ctor(ITestOutputHelper output) + 0x48
   at System.Private.Xml.Tests!<BaseAddress>+0x12ad20d
   at System.Reflection.DynamicInvokeInfo.Invoke(Object thisPtr, IntPtr methodToCall, Object[] parameters, BinderBundle binderBundle, Boolean wrapInTargetInvocationException) + 0x150
```

mdh1418 committed:
[tests] Condition problematic XmlSchemaValidatorApiTests classes on ReflectionEmit

mdh1418 commented:
/azp run runtime-extra-platforms

jkotas commented:
This is not fixing the whole problem, but it is still an improvement. Merging.

jkotas committed:
Revert "[tests] Condition problematic XmlSchemaValidatorApiTests classes on ReflectionEmit"

This reverts commit 103d59617fd0b12a31330e32895e19c13fc43924.

jkotas committed:
Revert "[tests] Condition XmlSerializerTests on ReflectionEmit"

This reverts commit 835d515bdd8c0b8eae02cc1c3ae98c30b23cbada.

jkotas commented:
Actually, I do not think that the condition on the Schema and XmlSerializer test is right (it should not be ReflectionEmit). The Xslt one is good, so let's merge that one at least.
